src/storage/mvcc/reader/point_getter.rs

```diff
@@ -0,0 +1,787 @@
+    drained: bool,
+}
+
+impl<S: Snapshot> PointGetter<S> {
+    /// Take out and reset the statistics collected so far.
+    #[inline]
+    pub fn take_statistics(&mut self) -> Statistics {
+        ::std::mem::replace(&mut self.statistics, Statistics::default())
+    }
+
+    /// Get the value of a user key.
+    ///
+    /// If `multi == false`, this function must be called only once. Future calls return nothing.
+    /// If `multi == true`, keys must be given in non-descending order. Calls with smaller keys
+    /// return nothing.
+    pub fn get(&mut self, user_key: &Key) -> Result<Option<Value>> {
+        if !self.multi {
+            // Protect from calling `get()` multiple times when `multi == false`.
+            if self.drained {
+                return Ok(None);
+            } else {
+                self.drained = true;
+            }
+        }
+
+        let mut ts = self.ts;
+
+        match self.isolation_level {
+            IsolationLevel::Si => {
+                // Check for locks that signal concurrent writes in Si.
+                match self.load_and_check_lock(user_key, ts)? {
+                    CheckLockResult::NotLocked => {}
+                    CheckLockResult::Locked(e) => return Err(e),
+                    CheckLockResult::Ignored(new_ts) => ts = new_ts,
+                }
+            }
+            IsolationLevel::Rc => {}
+        }
+
+        self.load_data(user_key, ts)
+    }
+
+    /// Get a lock of a user key in the lock CF. If lock exists, it will be checked to
+    /// see whether it conflicts with the given `ts`. If there is no conflict or no lock,
+    /// the safe `ts` will be returned.
+    #[inline]
+    fn load_and_check_lock(&mut self, user_key: &Key, ts: u64) -> Result<CheckLockResult> {
+        if self.multi {
+            self.load_and_check_lock_multi_get(user_key, ts)
+        } else {
+            self.load_and_check_lock_single_get(user_key, ts)
+        }
+    }
+
+    /// If only one `get()` will be called, we can use `snapshot.get_cf()` to get lock directly.
+    fn load_and_check_lock_single_get(
+        &mut self,
+        user_key: &Key,
+        ts: u64,
+    ) -> Result<CheckLockResult> {
+        self.statistics.lock.get += 1;
+        let lock_value = self.snapshot.get_cf(CF_LOCK, user_key)?;
+
+        if let Some(ref lock_value) = lock_value {
+            self.statistics.lock.processed += 1;
+            let lock = Lock::parse(lock_value)?;
+            super::util::check_lock(user_key, ts, &lock)
+        } else {
+            Ok(CheckLockResult::NotLocked)
+        }
+    }
+
+    /// If multiple `get()` will be called, we need to use cursor to read the lock.
+    fn load_and_check_lock_multi_get(
+        &mut self,
+        user_key: &Key,
+        ts: u64,
+    ) -> Result<CheckLockResult> {
+        self.ensure_lock_cursor()?;
+        let lock_cursor = self.lock_cursor.as_mut().unwrap();
+        if !self.lock_cursor_valid {
```

Comment:
Does it assume keys in multi-get is ordered?

Reply:
It's sorted here. https://github.com/tikv/tikv/pull/5424/files#diff-ef870876e008305d5f91928107ab70f9R170

Reply:
I know. I think it's better to note its assumption in PointGetter to prevent somebody uses it to get unsorted keys.

Reply:
It's also commented: https://github.com/tikv/tikv/pull/5424/files#diff-541556327b551ebbb7f586e06e03e81aR140

Reply:
It's a good idea to check it, but I'm more afraid of the overhead.

Reply:
> It's also commented: https://github.com/tikv/tikv/pull/5424/files#diff-541556327b551ebbb7f586e06e03e81aR140

🤣 I didn't notice it.

Reply:
@MyonKeminta Smaller keys should always returns `None` because the cursor will not move back.